### PR TITLE
handle non-overlap keys in hashFullOuterJoin, fix #2109

### LINKED ISSUE #2109: hashFullOuterJoin does not include items from right only collection
The following code demonstrates issue:
```scala
import com.spotify.scio._
val sc = ScioContext.forTest()
val left = sc.parallelize(Seq("a" -> 1))
val right = sc.parallelize(Seq("b" -> 2))
val hashJoined = left.hashFullOuterJoin(right).debug(prefix = "HashJoined: ")
val shuffleJoined = left.fullOuterJoin(right).debug(prefix = "ShuffleJoined: ")
sc.close().waitUntilDone()
```
The expected result - Shuffle and Hash should give exactly the same result - items from both collections:
```
ShuffleJoined: (b,(None,Some(2)))
ShuffleJoined: (a,(Some(1),None))
HashJoined: (b,(None,Some(2)))
HashJoined: (a,(Some(1),None))
```
Actual result - hash join does not return an item from `right` collection:
```
ShuffleJoined: (b,(None,Some(2)))
ShuffleJoined: (a,(Some(1),None))
HashJoined: (a,(Some(1),None))
```

Affected versions:
At least 0.7.4 and 0.8.0-alpha2.

### PR DESCRIPTION
**Pull request by nevillelyh**
Summary:

Beam `Combine.Globally` transform doesn't allow defaults when windowing is not the default `GlobalWindow`, and we erroneously added `.withoutDefaults()` to all `Combine.Globally` instances in `PCollectionWrapper`.

To fix properly, `.withoutDefaults()` should only be set when:
- `CombineFn` doesn't have a logical default, e.g. `reduce`, `Semigroup`, `Aggregator`, or
- `PCollection#getWindowStrategy != WindowingStrategy.globalDefault()`

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala [hunk 1]
```diff
@@ -42,7 +42,8 @@ private[values] trait PCollectionWrapper[T] extends TransformNameable {
     transform: PTransform[_ >: PCollection[T], PCollection[U]]
   ): SCollection[U] = {
     val t =
-      if (classOf[Combine.Globally[T, U]] isAssignableFrom transform.getClass) {
+      if ((classOf[Combine.Globally[T, U]] isAssignableFrom transform.getClass)
+          && internal.getWindowingStrategy != WindowingStrategy.globalDefault()) {
```
Comment (ClaireMcGinty) [inline scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala]: What's the reasoning behind modifying this `PTransform` in Scio, instead of letting Beam throw an error upon [expansion of the Combine.Globally transform](https://github.com/apache/beam/blob/c2f0d282337f3ae0196a7717712396a5a41fdde1/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java#L1184)? Since we're doing this modification "behind the scenes" users might get confused why their code isn't behaving like they'd expect?

Comment (nevillelyh) [reply scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala]: Without this, all Scio "global combine" methods will fail when there's windowing and there's no way to fix without dropping to Beam API.

File: scio-test/src/test/scala/com/spotify/scio/ArraySemigroupTest.scala [hunk 2]
```diff
@@ -34,25 +33,17 @@ class ArrayMonoidTest extends FlatSpec with Matchers {
   "ArrayMonoid" should "support plus" in {
```
Comment (ClaireMcGinty) [inline scio-test/src/test/scala/com/spotify/scio/ArraySemigroupTest.scala]: `ArrayMonoid` -> `ArraySemigroup`?

File: scio-core/src/main/scala/com/spotify/scio/package.scala [hunk 3]
```diff
@@ -30,20 +30,20 @@ import scala.reflect.ClassTag
  */
 package object scio {
 
-  /** [[com.twitter.algebird.Monoid Monoid]] for `Array[Int]`. */
-  implicit val intArrayMon: Monoid[Array[Int]] = new ArrayMonoid[Int]
+  /** [[com.twitter.algebird.Semigroup Semigroup]] for `Array[Int]`. */
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/package.scala]: tbh i don't like the fact that this implicits are here with this scope.

Given that they are, this is a breaking change. Is this change really necessary for this PR? Why changing from Monoid to Semigroup?

Comment (nevillelyh) [reply scio-core/src/main/scala/com/spotify/scio/package.scala]: `Array.empty` is not the same as array of fixed size. And there's no way to produce a correct `zeroValue` without seeing an array first.

Removing it from the scope is also a breaking change but probably lesser of 2 evils, assuming not many does `.sum` on arrays.

File: scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala (hunk 1 above)
Comment (ClaireMcGinty) [reply scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala]: cool, I was forgetting that user code doesn't directly interact directly with `pApply`. in that case isn't it redundant to add `.withoutDefaults())` to global combines elsewhere in this PR?

File: scio-core/src/main/scala/com/spotify/scio/package.scala (hunk 3 above)
Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/package.scala]: hm, doesn't it only affect the allocation? 

yeah, having this like this leads to that. However, i would probably push this change to another PR.

File: scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala (hunk 1 above)
Comment (nevillelyh) [reply scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala]: Hmm... They're not exactly the same, depending on windowing.

With windowing:
- `withoutDefaults` always needed, regardless whether the `fn` supports default value or not.

Without windowing:
- `aggregate(Aggregator)`, `combine`, `sum` - always need `withoutDefaults`, otherwise the runner might attempt to get default value and fail the assertion below.
`at com.spotify.scio.util.Functions$$anon$2.extractOutput(Functions.scala:171)`
- `fold`, `aggregate(zeroValue)` provides defaults and should not call `withoutDefaults` unless if there's windowing.

So basically it depends on the type of `CombineFn` which we can't determine here.

File: scio-core/src/main/scala/com/spotify/scio/package.scala (hunk 3 above)
Comment (nevillelyh) [reply scio-core/src/main/scala/com/spotify/scio/package.scala]: Downstream transform will make assumption about the array dimension, i.e. if you're `sum`ming 40 dimension vectors it'll assume the single output has size 40 and AIOB if it's 0.

Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/package.scala]: I think that's somewhat application specific and I it would need a different Monoid instance. This one is a lawful one (as long the array is not mutated) but also the simplest.

I think it will be better to mark these Monoid's as deprecated and then let users defined their instance according to their needs.

Comment (nevillelyh) [reply scio-core/src/main/scala/com/spotify/scio/package.scala]: Empty zero is not a lawful monoid, because  `zero + v` doesn't work if `v` has > 0 dimension and `+` is element-wise plus.
